src/mem_config.rs

```diff
@@ -0,0 +1,431 @@
+#![allow(clippy::needless_range_loop)]
+use num_integer::Roots;
+
+// Note: we do *not* need checked math here, because in all practical applications we will see
+//       out-of-memory long before the math overflows
+
+// The math here works with any non-zero value for FACTOR, but using powers of two should
+// optimize to significantly more efficient code.
+
+// IDEA: Some MemConfig methods can be somewhat expensive, iterators and other methods may use
+//       a cache, when the next indexing is within the same segment then no expensive
+//       calculations have to be done.
+
+/// Configures the sizes of segments and how to index entries. `Linear`, `Proportional` and
+/// `Exponential` implement this trait.
+pub trait MemConfig {
+    /// Called by ctors to assert that the configuration is valid.
+    ///
+    /// Some `MemConfig` implementations may put constraints on (const generic)
+    /// parameters. Currently it is impossible to assert these at compile time in stable
+    /// rust. In debug builds we check these constraints when a `SegVec` uses some
+    /// config. The three shipped `MemConfig` implementations check here that the 'FACTOR' is
+    /// not zero.
+    fn debug_assert_config();
```

Comment:
I think there are some stable ways to assert things about const generics now.

This is one example (stolen from a thread in an issue on static_assertions):

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=98221c637c7195f1fc530cc51d3848f8

If you feel up to it, I think it would be better to have these assertions happen at compile time... especially given that the (current) assertions are simple, just checking the FACTOR value against 0.